Add url error handling for the CheckResultDetailPage

EMaksy opened the pull request:
# Description

In the Check Detail Page served at  `/clusters/:clusterID/executions/last/:checkID/:targetType/:targetName` the combination of checkID, targetType, targetName and clusterID determines the information that is shown.

An URL with wrong parameters would still render the CheckResultDetailPage component.
This behavior is unwanted and should render a NotFound page. After discussing if the NotFound page should be rendered as a full view, we decided to render it inside for improved UI.

To handle this cases, follow changes were made:

* Added URL validation for CheckResultDetailPage.
* Refactored the NotFound component, it accepts parameters, in order to change the button text and navigation path.


## Demo

### Invalid clusterID
[invalid_clusterID.webm](https://user-images.githubusercontent.com/54111255/234282012-a7aa0ab3-bb19-41be-93e3-a35ec4ca70e4.webm)



### Invalid checkID, targetType, targetName
[invalidUrlParams.webm](https://user-images.githubusercontent.com/54111255/234282117-33361306-2b12-4e08-8b0f-28e72f887409.webm)


### NotFound story
![NotFoundStorybook](https://user-images.githubusercontent.com/54111255/234869834-d3dc44bc-7e41-41aa-b570-4ba87fb07c52.png)



## How was this tested?
* Added NotFound component test
* Added NotFound story
* Added CheckResultDetailPage test

dottorblaster commented on assets/js/components/NotFound/NotFound.test.jsx:
```diff
@@ -0,0 +1,47 @@
+
+import NotFound from '.';
+
+jest.mock('react-router-dom', () => ({
+  useNavigate: jest.fn(),
+}));
```
```suggestion

```

dottorblaster commented on assets/js/components/NotFound/NotFound.test.jsx:
```diff
@@ -0,0 +1,47 @@
+describe('NotFound', () => {
+  it('should render NotFound component correctly with default parameter values', () => {
+    const navigate = jest.fn();
+    useNavigate.mockReturnValue(navigate);
```
```suggestion

```

EMaksy replied:
Ah thanks! Not only this can go, but also we don't need  `const navigate = jest.fn();` and we can use render instead of `renderWithRouter` as NotFound page does not require `useNavigate ` mocks any more :D

dottorblaster commented on assets/js/components/NotFound/NotFound.test.jsx:
```diff
@@ -0,0 +1,47 @@
+  it('should render NotFound correctly with passed props', () => {
+    const url = faker.internet.url();
+    const buttonText = faker.lorem.word();
+    const navigate = jest.fn();
+    useNavigate.mockReturnValue(navigate);
```
```suggestion
const navigate = jest.fn();
```

dottorblaster commented on assets/js/components/ExecutionResults/CheckResultDetail/CheckResultDetailPage.test.jsx:
```diff
@@ -0,0 +1,166 @@
+    );
+    renderWithRouterMatch(StatefulCheckResultDetailPage, {
+      path: 'clusters/:clusterID/executions/last/:checkID/:targetType/:targetName',
+      route: `/clusters/${reduxStore.clustersList.clusters[0].id}/executions/last/${reduxStore.catalog.data[0].id}/${invalidTargetType}/${reduxStore.hostsList.hosts[0].hostname}`,
```
I think `reduxStore.clustersList.clusters[0].id` can be declared as a constant. Same goes for other variables

EMaksy replied:
Fair point, i wrote a helper function for that!

dottorblaster commented on assets/js/components/ExecutionResults/CheckResultDetail/CheckResultDetailPage.jsx:
```diff
@@ -13,8 +17,10 @@ import {
 } from '@state/lastExecutions';
 import { getHostID } from '@state/selectors/cluster';
 import ExecutionContainer from '@components/ExecutionResults/ExecutionContainer';
+import NotFound from '@components/NotFound/NotFound';
```
```suggestion
import NotFound from '@components/NotFound';
```

EMaksy replied:
Thanks!

dottorblaster commented on assets/js/components/NotFound/NotFound.stories.jsx:
```diff
@@ -0,0 +1,13 @@
+export default {
+  title: 'NotFound',
+  component: NotFound,
+  args: { buttonText: 'Go back home' },
```
```suggestion
  args: { buttonText: 'Go back home', onNavigate: () => {} },
```

EMaksy replied:
Nice catch thanks, as the story was written before the refactor xD

dottorblaster reviewed (changes requested):
Some little changes then we can merge

nelsonkopliku reviewed (commented):
Great job Eugen! Thanks for taking care of this.
The behavior looks good. Let's hear from @jagabomb for any suggestion about UX/UI.

Besides what already commented I believe that this work opens the possibility to improve and streamline error handling throughout the whole application.
However, let's defer further refactoring in this regards as it would require some deciasions/agreements to be made.

jagabomb commented:
> Great job Eugen! Thanks for taking care of this. The behavior looks good. Let's hear from @jagabomb for any suggestion about UX/UI.

@nelsonkopliku @EMaksy I think this solves the ask for displaying the a **clear 404 page** while retaining the sidebar for navigation purposes. The only outstanding issue I have is the alignment of elements on the 404 page. The "Go back home" button is touching the Trento logo, ideally there should be about a 40px space between this button and the logo.

EMaksy commented on assets/js/components/ExecutionResults/CheckResultDetail/CheckResultDetailPage.jsx:
```diff
@@ -25,8 +31,12 @@ import CheckDetailHeader from './CheckDetailHeader';
   const { clusterID, checkID, targetType, targetName } = useParams();
-
   const dispatch = useDispatch();
+  const navigateToUrl = useNavigate();
+
+  function onNavigate(path) {
+    navigateToUrl(path);
+  }
```
oh yeah thanks!

EMaksy commented on assets/js/lib/test-utils/factories/executions.js:
```diff
@@ -131,10 +131,21 @@ export const checksExecutionCompletedForTargetsFactory = Factory.define(
-      agents_check_results: targets.map(checkResultForTarget),
-    });
 
+    const checkResults = params.check_id
+      ? [
+          checkResultFactory.build({
+            agents_check_results: targets.map(checkResultForTarget),
+            check_id: params.check_id[0],
+          }),
+          checkResultFactory.build({
+            agents_check_results: targets.map(checkResultForTarget),
+            check_id: params.check_id[1],
+          }),
+        ]
```
Oh yeah, i need to adjust that it does not matter how many check_ids we pass :+1:

EMaksy commented:
Thanks team for the review, I addressed the comments!
@jagabomb  i added like you wished 40 px space between the logo and the button. (Screenshot is updated).

I'm with you team, the first step is done, we can discuss further error handling to improve even more :+1: .

dottorblaster reviewed (approved):
LGTM

nelsonkopliku reviewed (approved):
:rocket:
